### PR TITLE
fix: validate username before registering user

### PR DESCRIPTION
**Pull request by ricardogarim**
As per [CORE-544](https://rocketchat.atlassian.net/browse/CORE-544) and as discussed internally at [#rocketchat-dev](https://open.rocket.chat/channel/rocketchat-dev?msg=zGLpHpcKHLAJ65WBL), creating a new user with an invalid username (containing special characters) results in an error message, but the user is still created. This leads to an error when attempting to register again using the same email, as it is already registered.

To address this issue, I introduced a new method to validate usernames before registering users, ensuring that invalid usernames prevent user creation.

https://github.com/RocketChat/Rocket.Chat/assets/9621276/c6fa6408-bde5-4cf4-9c25-701b47d9a796

[CORE-544]: https://rocketchat.atlassian.net/browse/CORE-544?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apps/meteor/app/lib/server/functions/validateUsername.ts [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import { settings } from '../../../settings/server';
+
+export const validateUsername = (username: string): boolean => {
+	const settingsRegExp = settings.get('UTF8_User_Names_Validation');
+
+	const usernameRegExp = settingsRegExp ? new RegExp(`^${settingsRegExp}$`) : new RegExp('^[0-9a-zA-Z-_.]+$');
```
Comment (MarcosSpessatto) [inline apps/meteor/app/lib/server/functions/validateUsername.ts:6]: I think an invalid regex could throw an error here, right?

File: apps/meteor/app/lib/server/functions/validateUsername.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import { settings } from '../../../settings/server';
+
+export const validateUsername = (username: string): boolean => {
```
Comment (MarcosSpessatto) [inline apps/meteor/app/lib/server/functions/validateUsername.ts:3]: Can we create a unit test for this?

**Review by MarcosSpessatto — CHANGES_REQUESTED** [review]
I think there are more places that require this change to be applied (or at least reviewed)...

https://github.com/RocketChat/Rocket.Chat/blob/develop/apps/meteor/server/methods/registerUser.ts#L31 is one of them, but I think we should revisit all the places where users are being created, like LDAP, SAML, admin panel, etc.

File: apps/meteor/tests/unit/app/lib/server/functions/validateUsername.spec.ts [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+import proxyquire from 'proxyquire';
+
+const proxySettings = {
+	settings: {
+		get: (key: string): string | null => {
+			if (key === 'UTF8_User_Names_Validation') {
+				// Default value set at apps/meteor/server/settings/general.ts
+				return '[0-9a-zA-Z-_.]+';
+			}
+			return null;
```
Comment (KevLehman) [inline apps/meteor/tests/unit/app/lib/server/functions/validateUsername.spec.ts:11]: You can declare a stub for this (with sinon) and then use `stub.withArgs('UTF...').returns('')`

File: apps/meteor/tests/unit/app/lib/server/functions/validateUsername.spec.ts [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+		it('should return true for a username with allowed special characters', () => {
+			const result = validateUsername('username-_.');
+			expect(result).to.be.true;
+		});
+	});
+
+	describe('with custom regex settings', () => {
+		beforeEach(() => {
+			proxySettings.settings.get = (key: string) => {
```
Comment (KevLehman) [inline apps/meteor/tests/unit/app/lib/server/functions/validateUsername.spec.ts:50]: Use an stub

**Review by KevLehman — CHANGES_REQUESTED** [review]
add a changeset

**Review by KevLehman — APPROVED** [review]

**Review by MarcosSpessatto — APPROVED** [review]

**Pull request merged**